tools/sched_ext/scx_serialise.bpf.c

```diff
@@ -375,57 +360,114 @@ static u32 get_executor_id(const struct task_struct *p) {
+		return -1;
+	}
+
+	tctx->priority = priority;
+	tctx->enqueued = true;
+	bpf_spin_unlock(&tctx->lock);
+	return 0;
+}
+
+static long update_priority_callback(struct bpf_map *map, const pid_t *pid, struct task_ctx *tctx, const u32 *eid) {
+	if (tctx->eid == *eid) {
+		if (update_priority(*pid, tctx) < 0) {
+			bpf_printk("[update_priority_callback] Failed to update priority for pid: %d\n", *pid);
+			return -1;
+		}
+	}
+	return 0;
+}
+
+/* 
+ * This function handles tasks flagged with SCHED_EXT to trigger concurrency issues.
+ * 
+ * - First, it checks for new scheduling requests from the user space via a ring buffer. If new requests 
+ *   are found, it add the request to `sched_job_map` and initializes the scheduling algorithm.
+ * 
+ * - It then looks up the task's executor ID (`eid`) and updates the scheduling job status for 
+ *   that executor in the `sched_job_map`. The tasks to be scheduled then follows asynchronously
+ *   and the function should be invoked repeatedly. Suppose the following tasks are desired.
+ * 
+ * - If the task is newly created and not yet in the `task_ctx_map`, it creates a new task context 
+ *   and adds it to the map. Keep waiting for other tasks in from the same executor.
+ * 
+ * - Finally, if all tasks for the given executor are ready, it updates task priorities and enqueues 
+ *   the `eid` for dispatch in the `group_dispatch_queue`.
+ */
+static void handle_sched_ext(struct task_struct *p)
+{
+    if (bpf_user_ringbuf_drain(&user_ringbuf, receive_req, NULL, 0) > 0) {
+        init_scheduling_algo();
+        struct event *e = bpf_ringbuf_reserve(&kernel_ringbuf, sizeof(*e), 0);
+        if (e) {
+            e->pid = p->pid;
+            bpf_printk("[handle_sched_ext] pid: %d\n", p->pid);
+            bpf_ringbuf_submit(e, 0);
+        }
+    }
+
+    u32 eid = get_executor_id(p);
+	// Lookup the scheduling job for this executor ID in the job map.
+    struct sched_job *job = bpf_map_lookup_elem(&sched_job_map, &eid);
+    if (!job) {
+        bpf_printk("[handle_sched_ext] job not found\n");
+        return;
+    }
+    job->num_ready++;
+	if (bpf_map_update_elem(&sched_job_map, &eid, job, BPF_EXIST) < 0) {
+        bpf_printk("[handle_sched_ext] fail to update job\n");
+		return;   
+	}
+
+	struct task_ctx* tctx = bpf_map_lookup_elem(&task_ctx_map, &p->pid);
+	if (!tctx) {
+		// If it is a newly created task, add it into `task_ctx_map`
+		struct task_ctx new_task = {
+ 			.priority = 0,
+		    .eid = eid,
+		    .enqueued = false,
+    		.lock = {},
+		};
+		if (bpf_map_update_elem(&task_ctx_map, &p->pid, &new_task, BPF_NOEXIST) < 0) {
+			bpf_printk("[handle_sched_ext] fail to insert new task ctx\n");
+			return;
+		}
+	}
+
+	int num_expected = (!tctx ? job->num_total:job->num_alive);
+	// If all the tasks are ready, update task priorities.
+	bool all_tasks_ready = (job->num_ready == num_expected);
+
+	if (all_tasks_ready) {
+    	bpf_for_each_map_elem(&task_ctx_map, update_priority_callback, &eid, 0);
```

Comment:
updating here is racy -- priorities need to be updated *before* tasks are marked as ready.

Dispatches calls are happening all the time asynchronously, so the following is possible

1. The tasks might all be ready
2. Dispatch is called and the next task is scheduled
3. Tasks are all assigned priorities (but initial task was already dispatched without a priority)